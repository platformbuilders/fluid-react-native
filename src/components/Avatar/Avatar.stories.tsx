import type { Meta } from '@storybook/react';
import Avatar from '.';

const meta: Meta<typeof Avatar> = {
  title: 'Avatar',
  component: Avatar,
  argTypes: { onPress: { action: 'clicked' } },
  args: {
    id: 'avatar-id',
    accessibility: 'Avatar',
    accessibilityLabel: 'Avatar',
    borderColor: '#355deb',
    borderWidth: 4,
    image: { uri: 'https://thispersondoesnotexist.com', cache: 'cacheOnly' },
    onPress: () => {},
    size: 220,
    testID: 'avatar-test-id',
  },
  parameters: {
    notes:
      'Esse componente tem como função auxiliar na manipulação, manutenção e padronização de avatares dentro de uma aplicação.',
  },
};

export default meta;

export const Default = {};
