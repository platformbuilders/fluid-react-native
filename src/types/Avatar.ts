export type AvatarType = {
  ref?: any;
  image?: string;
  name?: string;
  showBorder?: boolean;
  borderWidth?: number;
  borderColor?: string;
  displayCamera?: boolean;
  size?: number;
  onPress?: (x: any) => void;
  onUpload?: (x: any) => any;
  id?: string;
  accessibility: string;
  accessibilityLabel?: string;
  testID?: string;
};
