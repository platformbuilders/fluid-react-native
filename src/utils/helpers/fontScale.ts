import { RFValue } from 'react-native-responsive-fontsize';

export const scaledFontSize = (
  size: number | string,
  baselineFactor?: 280,
): number => {
  const value = typeof size === 'string' ? parseInt(size, 10) : size;
  return RFValue(value, baselineFactor);
};
