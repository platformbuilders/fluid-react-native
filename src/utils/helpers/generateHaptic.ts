import Haptic, { HapticFeedbackType } from 'react-native-haptic';
import { isIOS } from './isIOS';

export const generateHaptic = (haptic: HapticFeedbackType = 'impact'): void => {
  if (isIOS()) {
    Haptic.generate(haptic);
  }
};
