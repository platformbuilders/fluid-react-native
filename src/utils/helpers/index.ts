export {
  currencyParser,
  parseToThousands,
  removeWhiteSpaces,
  toOnlyNumbers,
  getTheme,
  ifStyle,
  switchStyle,
} from '@platformbuilders/helpers';
export * from './generateHaptic';
export * from './getByTypographyVariant';
export * from './fontScale';
export * from './shadow';
