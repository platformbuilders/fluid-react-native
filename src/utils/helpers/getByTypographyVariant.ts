import { getTheme } from '@platformbuilders/helpers';
import { scaledFontSize } from './fontScale';

export const getFontSize = (props: any): number => {
  const fontSize = getTheme(`${props.variant}.fontSize`)(props) || 16;
  return scaledFontSize(fontSize);
};

export const getLineHeight = (props: any): number => {
  const lineHeight = getTheme(`${props.variant}.lineHeight`)(props) || 22;
  return scaledFontSize(lineHeight);
};
