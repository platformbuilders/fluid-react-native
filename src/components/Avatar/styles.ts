import styled from 'styled-components/native';
import { RNCamera } from 'react-native-camera';
import Touchable from '../Touchable';
import Icon from '../Icon';
import { getTheme, ifStyle } from '../../utils/helpers';

const accentMain = getTheme('accent.main');
const accentContrast = getTheme('accent.contrast');

const showBorder = ifStyle('showBorder');
const hasBorderWidth = ifStyle('borderWidth');
const hasBorderColor = ifStyle('borderColor');

type WrapperProps = {
  size: number;
  disabled: boolean;
  showBorder: boolean;
  borderWidth?: number;
  borderColor?: string;
};

export const Wrapper = styled(Touchable)<WrapperProps>`
  width: ${(props: WrapperProps): number => props.size}px;
  height: ${(props: WrapperProps): number => props.size}px;
  border-radius: ${(props: WrapperProps): number => props.size / 2}px;
  overflow: hidden;
  border-width: ${(props: WrapperProps): any =>
    showBorder(hasBorderWidth(props.borderWidth, '2'), '0')}px;
  border-color: ${(props: WrapperProps): any =>
    hasBorderColor(props.borderColor, accentMain)};
  position: relative;
`;

type IconWrapperProps = {
  size: number;
};

export const UploadIconWrapper = styled.View<IconWrapperProps>`
  align-items: center;
  justify-content: center;
  width: ${(props: IconWrapperProps): number => props.size / 4}px;
  height: ${(props: IconWrapperProps): number => props.size / 4}px;
  border-radius: ${(props: IconWrapperProps): number => props.size / 8}px;
  position: absolute;
  z-index: 2;
  background-color: ${accentMain};
`;

export const UploadIcon = styled(Icon).attrs({ name: 'camera' })`
  color: ${accentContrast};
`;

export const CameraView = styled(RNCamera)<{ size: number }>`
  width: ${(props: any): number => props.size}px;
  height: 100%;
  border-radius: ${(props: any): number => props.size / 2}px;
  overflow: hidden;
  border: ${showBorder('4px solid white', '')};
`;
