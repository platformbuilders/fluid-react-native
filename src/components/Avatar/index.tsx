import React, { useState, useRef, useImperativeHandle } from 'react';
import { isEmpty } from 'lodash';
import { RNCamera } from 'react-native-camera';
import FastImage, { Source } from 'react-native-fast-image';
import ImagePicker from 'react-native-image-picker';
import If from '../If';
import { ImageAvatarPlaceholder as defaultAvatar } from '../../assets/images';
import { Wrapper, CameraView, UploadIconWrapper, UploadIcon } from './styles';

type Props = {
  ref?: any;
  image?: string;
  showBorder?: boolean;
  borderWidth?: number;
  borderColor?: string;
  displayCamera?: boolean;
  size?: number;
  onPress?: (x: any) => void;
  onUpload?: (x: any) => any;
  id?: string;
  accessibility: string;
  accessibilityLabel?: string;
  testID?: string;
};

const Avatar: React.FC<Props> = React.forwardRef(
  (
    {
      id,
      image = defaultAvatar,
      accessibility = 'Upload de Avatar',
      accessibilityLabel,
      testID,
      size = 50,
      onPress,
      onUpload,
      showBorder = true,
      displayCamera = false,
      ...rest
    },
    ref,
    // eslint-disable-next-line sonarjs/cognitive-complexity
  ) => {
    const [uploadedImage, setUploadedImage] = useState<any>();
    const cameraRef = useRef<any>();

    const openPicker = (): Promise<void> => {
      const options = {
        title: 'Selecionar foto',
        cancelButtonTitle: 'Cancelar',
        takePhotoButtonTitle: 'Tirar foto',
        chooseFromLibraryButtonTitle: 'Escolher da galeria',
        quality: 0.3,
        storageOptions: {
          skipBackup: true,
          path: 'images',
        },
      };
      return new Promise((resolve) => {
        ImagePicker.launchImageLibrary(options as any, (response: any) => {
          setUploadedImage(response.uri);
          if (onUpload) {
            onUpload(response.uri);
          }
          resolve();
        });
      });
    };

    const takePicture = async (): Promise<void> => {
      if (cameraRef.current) {
        const options = { quality: 0.5, base64: true };
        const data = await cameraRef.current.takePictureAsync(options);
        setUploadedImage(data);
        return data;
      }
    };

    const getUploadImage = (): any => {
      return uploadedImage;
    };

    const clearUploadImage = (): void => {
      setUploadedImage('');
    };

    const getCurrentAvatar = (): Source | any => {
      if (uploadedImage) {
        return { uri: uploadedImage };
      }
      if (image && !isEmpty(image)) {
        return { uri: image };
      }
      return defaultAvatar;
    };

    useImperativeHandle(ref, () => ({
      getUploadImage,
      clearUploadImage,
      takePicture,
      openPicker,
    }));

    return (
      <Wrapper
        id={id || accessibility}
        accessibility={accessibility}
        accessibilityLabel={accessibilityLabel || accessibility}
        testID={testID || id}
        size={size}
        onPress={onPress}
        disabled={!onPress}
        showBorder={showBorder}
        {...rest}
      >
        {displayCamera && !uploadedImage ? (
          <CameraView
            ref={cameraRef}
            size={size}
            type={RNCamera.Constants.Type.front}
            flashMode={RNCamera.Constants.FlashMode.auto}
            androidCameraPermissionOptions={{
              title: 'Câmera',
              message: 'Precisamos da sua permissão para usar a câmera',
              buttonPositive: 'Ok',
              buttonNegative: 'Cancelar',
            }}
          />
        ) : (
          <FastImage
            source={getCurrentAvatar()}
            resizeMode={FastImage.resizeMode.cover}
            style={{ width: '101%', height: '101%' }}
          />
        )}
        <If condition={!!displayCamera}>
          <UploadIconWrapper size={size}>
            <UploadIcon id="" accessibility="" />
          </UploadIconWrapper>
        </If>
      </Wrapper>
    );
  },
);

Avatar.displayName = 'AvatarComponent';

export default Avatar;
