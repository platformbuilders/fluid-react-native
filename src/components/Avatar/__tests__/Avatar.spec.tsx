import 'jest';
import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';
import { ThemeProvider } from 'styled-components/native';
import Avatar from '..';
import { theme } from '../../../test/helpers';

describe('<Avatar />', () => {
  it('should render Avatar', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <Avatar id="testing" accessibility="" />
      </ThemeProvider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render Avatar with camera', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <Avatar id="testing" accessibility="" displayCamera />
      </ThemeProvider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should trigger onPress function', () => {
    const onPressEvent = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Avatar id="testing" accessibility="" onPress={onPressEvent} />
      </ThemeProvider>,
    );

    const component = getByTestId('testing');

    fireEvent.press(component);

    expect(onPressEvent).toHaveBeenCalled();
  });
});
