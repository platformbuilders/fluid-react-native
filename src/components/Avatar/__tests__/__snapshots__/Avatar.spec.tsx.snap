// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> should render Avatar 1`] = `
<View
  accessibilityLabel=""
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#547E6A",
      "borderRadius": 25,
      "borderWidth": 2,
      "height": 50,
      "opacity": 1,
      "overflow": "hidden",
      "position": "relative",
      "width": 50,
    }
  }
  testID="testing"
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": "101%",
          "width": "101%",
        },
      ]
    }
  >
    <FastImageView
      resizeMode="cover"
      source={
        Object {
          "uri": Object {
            "testUri": "../../../src/assets/images/avatar-placeholder.png",
          },
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`<Avatar /> should render Avatar with camera 1`] = `
<View
  accessibilityLabel=""
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#547E6A",
      "borderRadius": 25,
      "borderWidth": 2,
      "height": 50,
      "opacity": 1,
      "overflow": "hidden",
      "position": "relative",
      "width": 50,
    }
  }
  testID="testing"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
  <View
    size={50}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#547E6A",
          "borderRadius": 6.25,
          "height": 12.5,
          "justifyContent": "center",
          "position": "absolute",
          "width": 12.5,
          "zIndex": 2,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "color": "#ffffff",
        }
      }
    >
      <View
        accessibilityLabel="Camera"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID=""
      >
        <View>
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ó°„€
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
